src/hooks/web3/index.tsx

```diff
@@ -0,0 +1,139 @@
+      web3ReactConnectors.walletConnect,
+      web3ReactConnectors.network,
+    ]
+    return prioritizedConnectors.filter((connector): connector is Web3ReactConnector => Boolean(connector))
+  }, [web3ReactConnectors])
+
+  const connectors = useMemo(
+    () => ({
+      user: web3ReactConnectors.user?.[0],
+      metaMask: web3ReactConnectors.metaMask[0],
+      walletConnect: web3ReactConnectors.walletConnect[0],
+      walletConnectQR: web3ReactConnectors.walletConnectQR[0],
+      network: web3ReactConnectors.network[0],
+    }),
+    [web3ReactConnectors]
+  )
+
+  // Attempt to connect eagerly if there is no user-provided provider.
+  useEffect(() => {
+    // Ignore any errors during connection so they do not propagate to the widget.
+    if (connectors.user) {
+      connectors.user.activate().catch(() => undefined)
+      return
+    }
+    for (const connector of [connectors.metaMask, connectors.walletConnect]) {
+      connector.connectEagerly().catch(() => undefined)
+    }
+    connectors.network.activate().catch(() => undefined)
+  }, [connectors.metaMask, connectors.network, connectors.user, connectors.walletConnect])
+
+  return (
+    <Web3ReactProvider connectors={prioritizedConnectors} key={key.current}>
+      <JsonRpcUrlMapProvider value={jsonRpcUrlMap}>
+        <ConnectorsProvider value={connectors}>{children}</ConnectorsProvider>
+      </JsonRpcUrlMapProvider>
+    </Web3ReactProvider>
+  )
+}
+
+const onError = (error: Error) => console.error(error)
+
+function initializeWeb3ReactConnector<T extends Connector, P extends object>(
+  Constructor: { new (options: P): T },
+  options?: Omit<P, 'actions'>
+): Web3ReactConnector<T> {
+  const [connector, hooks] = initializeConnector((actions) => new Constructor({ actions, onError, ...options } as P))
+  return [connector, hooks]
+}
+
+function useWeb3ReactConnectors(
+  defaultChainId: SupportedChainId,
+  jsonRpcUrlMap: JsonRpcUrlMap,
+  provider?: Eip1193Provider | JsonRpcProvider
+) {
+  const user = useMemo(() => {
+    if (!provider) return
+    if (JsonRpcProvider.isProvider(provider)) {
+      return initializeWeb3ReactConnector(JsonRpcConnector, { provider })
+    } else if (JsonRpcProvider.isProvider((provider as any).provider)) {
+      throw new Error('Eip1193Bridge is experimental: pass your ethers Provider directly')
+    } else {
+      return initializeWeb3ReactConnector(EIP1193, { provider })
+    }
+  }, [provider])
+  const metaMask = useMemo(() => initializeWeb3ReactConnector(MetaMask), [])
+  const walletConnect = useMemo(
+    () =>
+      initializeWeb3ReactConnector(WalletConnect, { options: { rpc: jsonRpcUrlMap, qrcode: true }, defaultChainId }),
+    [defaultChainId, jsonRpcUrlMap]
+  ) // WC via built-in popup
+  const walletConnectQR = useMemo(
+    () =>
+      initializeWeb3ReactConnector(WalletConnect, { options: { rpc: jsonRpcUrlMap, qrcode: false }, defaultChainId }),
+    [jsonRpcUrlMap, defaultChainId]
+  ) // WC via tile QR code scan
```

Comment:
maybe I'm reading this wrong, but it seems like the comments on these WalletConnect connector instances are reversed?

Reply:
These are confusingly named, and the qrcode property just makes it more confusing.
I'm leaving for now, but in the future we can look into cleaning this up by only using a single connector.